### DISCUSSION TITLE
Connecting to Teleport nodes with OpenSSH

### DISCUSSION THREAD

**Discussion by webvictim** [gravitational/teleport#5530] [Show and tell]
### Update (November 2022)

- `tsh config` can now automatically generate you an SSH configuration file to use with OpenSSH
- it uses `tsh proxy ssh` as a `ProxyCommand` so you don't need to think about arguments and building your own configuration to the same extent

It's worth trying to use `tsh config` before you invest all the time and effort from this post - but I'm leaving this here as a learning exercise.

# Guide

One question we get often is how you can use OpenSSH to easily connect to Teleport clusters. I'd like to share an example of the configurations I use to handle this personally on my home network.

I've created a `~/.ssh/teleport` directory to hold these files. I find it best to keep my Teleport SSH configurations away from my regular SSH configurations as it can be hard to dig out the correct `known_hosts` entries from a file with hundreds of lines in it. It also means I can check my Teleport configurations into git to keep them in sync across a variety of different hosts.

I use just one line in my `~/.ssh/config` file to include this separate configuration file:

`~/.ssh/config`:
```
Include ~/.ssh/teleport/ssh_config
```

Magic!

### 1) SSH config file

This file contains two entries - one for the proxy and one for a specific host. I'll explain what each line does.

`~/.ssh/teleport/ssh_config`:
```bash
# Proxy configuration
# This only needs to be in the file once, I put it at the top.
# The DNS name that I use to access my Teleport proxy
Host teleport.example.com
    # The UNIX login used to connect to the Teleport proxy - this must be one of the allowed
    # logins for your Teleport role/user
    User gus

    # The port where the Teleport SSH proxy is listening
    Port 3023

    # Path to the private key used to authenticate
    # SSH automatically expands this to also read `webvictim-cert.pub` in the same directory
    # (which contains the SSH certificate issued by Teleport)
    # You can build this path consistently using this information:
    # - ~/.tsh/keys is the default location where certificates are stored locally when logging in with tsh login
    # - the directory is the DNS name of the proxy (teleport.example.com)
    # - the filename is the name of the Teleport user I'm using to log in (webvictim)
    IdentityFile ~/.tsh/keys/teleport.example.com/webvictim 

    # (optional) Path to the SSH certificate used for authentication, needed for tsh 6.0+
    # Prior to tsh 6.0, this file lived at ~/.tsh/keys/teleport.example.com/webvictim-cert.pub and would be automatically loaded
    # After tsh 6.0 the path must be provided separately if there is no usable ssh-agent.
    # You can build this path consistently using this information:
    # - ~/.tsh/keys is the default location where certificates are stored locally when logging in with tsh login
    # - the first directory is the DNS name of the proxy (teleport.example.com)
    # - the second directory is the name of the Teleport user I'm using to log in, followed by -ssh (webvictim-ssh)
    # - the filename is the name of the Teleport cluster, followed by -cert.pub (teleport.example.com-cert.pub)
    CertificateFile ~/.tsh/keys/teleport.example.com/webvictim-ssh/teleport.example.com-cert.pub

    # The path to my SSH known hosts file which contains the fingerprint of my Teleport host CA
    # SSH uses this to validate that the host it's connecting to should be trusted
    # Having this set means that I never get a trust-on-first-use prompt from SSH asking me to answer
    # "yes" before logging into a host
    UserKnownHostsFile ~/.ssh/teleport/known_hosts

# Node configuration
# The hostname of the host I'm accessing, as displayed in the output of 'tsh ls'
# This is called when I run 'ssh zeus' at the command prompt
Host zeus
    # See 'User' section above
    User gus

    # A command which runs before connecting to the SSH host
    # We use this to:
    # 1) call a login script which checks that my Teleport certificates are valid and handles
    # reauthentication if they have expired
    # 2) establishes an SSH connection to the Teleport proxy which will pass traffic along to
    # my SSH host
    # The use of 'ssh teleport.example.com' means that SSH reads this config file again
    # and automatically applies the connection settings for 'teleport.example.com' listed above
    ProxyCommand sh -c '~/.ssh/teleport/login.sh teleport.example.com webvictim && ssh teleport.example.com -s proxy:%h:%p'

    # The port to connect to on the node, in this case it's 3022 for the Teleport node/SSH service
    Port 3022

    # See 'IdentityFile' above
    IdentityFile ~/.tsh/keys/teleport.example.com/webvictim

    # See 'CertificateFile' above
    CertificateFile ~/.tsh/keys/teleport.example.com/webvictim-ssh/teleport.example.com-cert.pub

    # When a node joins a Teleport cluster, Teleport automatically issues a certificate for it
    # which matches this pattern: <hostname>.<proxy address>
    # In my case, the hostname is 'zeus' and my proxy address is 'teleport.example.com',
    # so I know that 'zeus.teleport.example.com' will be one of the principals which Teleport
    # will add to the host certificate for my node.
    # This is relevant because when SSH connects to a host and a host certificate is presented,
    # it validates that the hostname is present on the certificate
    # If the hostname is not presented, you will get a trust-on-first-use prompt because it doesn't
    # match the host CA fingerprint listed in ~/.ssh/teleport/known_hosts
    # We want to avoid these prompts, so we use HostKeyAlias to tell SSH what hostname to expect
    # to match against.
    # If you're curious about what principals are on a host certificate being presented, you can check:
    # See the 'Checking host principals' section below.
    HostKeyAlias zeus.teleport.example.com

    # See 'UserKnownHostsFile' above
    UserKnownHostsFile ~/.ssh/teleport/known_hosts
```

### 2) Auto-login script

As you'll see above, there is a a script called `login.sh` which runs as part of my SSH `ProxyCommand`. This is that script:

`~/.ssh/teleport/login.sh`:
```bash
#!/bin/bash
TSH=/usr/local/bin/tsh
if [ ! -f ${TSH} ]; then
    echo "Cannot find tsh at '${TSH}'"
    exit 2
fi

if ! command -v ssh-keygen; then
    echo "Cannot find ssh-keygen"
    exit 2
fi

if [[ "$1" == "" ]]; then
    echo "Usage: $(basename $0) <proxy> <username>"
    exit 1
fi
PROXY=$1

if [[ "$2" == "" ]]; then
    echo "Usage: $(basename $0) <proxy> <username>"
    exit 2
fi
USERNAME=$2

CERT_FILE=""
# The file location changed in tsh ~v6
if [ -f ~/.tsh/keys/${PROXY}/${USERNAME}-ssh/${PROXY}-cert.pub ]; then
    CERT_FILE=~/.tsh/keys/${PROXY}/${USERNAME}-ssh/${PROXY}-cert.pub
# Older tsh file location
elif [ -f ~/.tsh/keys/${PROXY}/${USERNAME}-cert.pub ]; then
    CERT_FILE=~/.tsh/keys/${PROXY}/${USERNAME}-cert.pub
fi

if [[ "${CERT_FILE}" != "" && -f ${CERT_FILE} ]]; then
    export EXPIRY_DATE=$(ssh-keygen -Lf ${CERT_FILE} | grep Valid | cut -d' ' -f13)
    # the date command is different on OS X, of course
    if [[ $OSTYPE == darwin* ]]; then
        EXPIRY_TIME=$(date -j -f '%Y-%m-%dT%H:%M:%S' ${EXPIRY_DATE} +%s)
    elif [[ $OSTYPE == linux* ]]; then
        EXPIRY_TIME=$(date -d ${EXPIRY_DATE} +%s)
    else 
        echo "Unsupported OS '${OSTYPE}'"
        exit 2
    fi

    NOW=$(date +%s)
    if [ ${NOW} -lt ${EXPIRY_TIME} ]; then
        exit 0
    fi
fi
${TSH} login --proxy=${PROXY} --user=${USERNAME}
```

This script takes two command line arguments:
1) Teleport proxy DNS name
2) Teleport username to authenticate as

This means I call the script using `~/.ssh/teleport/login.sh teleport.example.com webvictim`. It will check whether my Teleport certificates have expired and if they have, it will run `tsh login` to reauthenticate.

Note: This script currently only works for SSO-based logins like Github, SAML or OIDC where a browser window is opened to handle authentication. It does not support providing a password for a Teleport user. In this situation, you may see an error like `inappropriate ioctl for device`. The workaround here is to run `tsh login` manually before trying to connect to a host via `ssh`.

### 3) Known hosts file

This is the final piece. This file contains the output of `tctl auth export --type=host` run on my Teleport auth server:

`~/.ssh/teleport/known_hosts`:
```console
@cert-authority teleport.example.com,*.teleport.example.com ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDX2iIvsuLDdAqd1Kd0LAKa+JRkF/qRNTCqrIuBqjV4F8i0lTQ+8XTr7dqoT1BrzEo39mn8SfjYWyULEGYPwtKr1VXkGhS7Oo6r/jabDuuKNecYFLycxZdbBXItGCOLcbmWwo+xWcHNzmtPa/SERrvsUepS0xsMQR+sOgCy11RQUgpLuy0LDah0toGEeAcmmyZXe0th/11z248GqvfOfGmExbeaVts1pJsSOvUqyO++S2VZmVs0dYiZTYWhu7QUha+yBPowd9X3zzI/gLuxUHat3heqcIfa9WYVwnY9pLD8J+b/ULciOXTwL4KKhk27SjmVqlnwmfWc2FoZc1wlCz/x type=host
```

(this information is a public key and is not sensitive - it does not need to be heavily guarded)

I have changed the part after `@cert-authority` to match two things with a comma separated list:
- the DNS name of my Teleport proxy
- `*.teleport.example.com` i.e. a wildcard for any certificate which is issued by my Teleport server
  - this works along with my `HostKeyAlias` entries above to tell SSH that it should always trust host certificates presented by Teleport hosts which have hostname principals on their certificates which match this pattern.

### Checking host principals

After you've seen all this mention of `HostKeyAlias` and host certificates, you may be wondering how to check what principals are being presented.

You can do this with `ssh-keyscan` and `ssh-keygen` - two utilities distributed along with the OpenSSH client.

A command like this will show you the host certificate being presented by your Teleport proxy: `ssh-keyscan -p 3023 -c teleport.example.com | ssh-keygen -Lf -`. The `-p 3023` connects to port 3023 (Teleport SSH proxy port) and `-c` instructs `ssh-keyscan` to look for host certificates rather than key fingerprints.

Here's some output:

```
$ ssh-keyscan -p 3023 -c teleport.example.com | ssh-keygen -Lf -
# teleport.example.com:3023 SSH-2.0-Teleport
# teleport.example.com:3023 SSH-2.0-Teleport
# teleport.example.com:3023 SSH-2.0-Teleport
# teleport.example.com:3023 SSH-2.0-Teleport
# teleport.example.com:3023 SSH-2.0-Teleport
(stdin):1:
        Type: ssh-rsa-cert-v01@openssh.com host certificate
        Public key: RSA-CERT SHA256:OVqmyvYy6zhmIDz+PEKxm5c/+pkEaiHZ9Bj1xzu5l+I
        Signing CA: RSA SHA256:1paZOuzvmf376q20ZZtfmedb0tS7yYGAGoetMh572aQ (using ssh-rsa)
        Key ID: ""
        Serial: 0
        Valid: after 2020-11-19T17:52:53
        Principals: 
                d1236170-0a87-4727-a151-8b7bc2441087.teleport.example.com
                d1236170-0a87-4727-a151-8b7bc2441087
                localhost
                127.0.0.1
                ::1
                artemis
                teleport.example.com
                remote.kube.proxy.teleport.cluster.local
        Critical Options: (none)
        Extensions: 
                x-teleport-authority UNKNOWN OPTION (len 23)
                x-teleport-role UNKNOWN OPTION (len 9)
```

Any of the entries under the `Principals` section could be used for a `HostKeyAlias`. You can see that `teleport.example.com` is present, which means SSH can correctly match the host certificate against the `@cert-authority` entry in `~/.ssh/teleport/known_hosts`.

You can also use `ssh-keyscan` to connect to Teleport nodes and check the principals which are presented - just use `-p 3022` instead when connecting to the nodes. One thing that's worth bearing in mind is that Teleport nodes connected to the proxy via [node tunnelling mode ](https://goteleport.com/teleport/docs/admin-guide/#adding-a-node-located-behind-nat) do not listen on port 3022 and must be connected to via the Teleport proxy, so you won't be able to check their principals using the `ssh-keyscan` method.

## Adding another host

When I need to add a new host, I copy the existing `Host zeus` section and update all the mentions of `zeus` to the new hostname. I can then `ssh newhost` completely transparently and everything just works as usual.

## Why go to all this effort?

The benefit of using `ssh` is that many, many tools already support it and most will automatically read a `~/.ssh/config` file if it's present. `scp`, `rsync` and other traditional Unix tools also just work as expected.

I have pretty strong muscle memory for typing `ssh host` to connect to hosts on my local network, so setting everything up this way means I never have to think about the fact that Teleport is there - it gets completely out of my way and just lets me use SSH as normal but I get all the benefits of Teleport - my sessions are recorded, all my access is going via a bastion host using short-lived user certificates and all my hosts are authenticated using their host certificates.

Let me know if you can think of improvements! How do you use SSH with Teleport yourself?

### Edit history

13th July 2021: Added `CertificateFIle` and details.
22nd November 2022: Added note about `tsh config`.

**Comment by webvictim** [comment]
@lamhoangtung I am replying to your comment from https://github.com/gravitational/teleport/issues/4176 here to keep everything in the same place.

```Certificate invalid: name is not a listed principal```

This error usually means that the hostname of the host you're connecting to does not appear in the principals listed on the certificate. You may need to use `HostKeyAlias` to make the hostname match the pattern associated with the host CA in your `~/.ssh/known_hosts` file.

You can try running `ssh -vv host` to get some more detailed debugging information. If you share a few lines before and after the `name is not a listed principal` error then I can try to help you figure it out.

**Reply by lamhoangtung** [reply]
@webvictim. Thanks for the help. Here are those fews line:

```console
debug2: Server host certificate hostname: 1983e6c6-6c27-4ea8-b05d-e69d2bb19cd2.teleport
debug2: Server host certificate hostname: 1983e6c6-6c27-4ea8-b05d-e69d2bb19cd2
debug2: Server host certificate hostname: techainer-rtx3090-1.teleport
debug2: Server host certificate hostname: techainer-rtx3090-1
debug2: Server host certificate hostname: localhost
debug2: Server host certificate hostname: 127.0.0.1
debug2: Server host certificate hostname: ::1
debug1: using hostkeyalias: techainer-rtx3090-1.tele.techainer.com
debug1: Host 'techainer-rtx3090-1.tele.techainer.com' is known and matches the RSA-CERT host certificate.
debug1: Found CA key in /Users/linus/.ssh/teleport/known_hosts:1
Certificate invalid: name is not a listed principal
debug1: No matching CA found. Retry with plain key
debug1: Host 'techainer-rtx3090-1.tele.techainer.com' is known and matches the RSA host key.
debug1: Found key in /Users/linus/.ssh/teleport/known_hosts:2
debug2: set_newkeys: mode 1
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug2: set_newkeys: mode 0
debug1: rekey in after 134217728 blocks
```

I though setting the principal in `~/.ssh/teleport/known_hosts` is enough as listed in the tutorial above. Also since I had SSH permission into the server that running Teleport, in my `~/.ssh/known_hosts` there is a principal for the same server (but in raw IP address, not domain). Is that going to be a problem ?

**Reply by webvictim** [reply]
So your issue here is that the HostKeyAlias `techainer-rtx3090-1.tele.techainer.com` does not appear as one of the entries for `Server host certificate hostname`. If you set HostKeyAlias to `techainer-rtx3090-1.teleport` it should work. Basically SSH needs to be able to do a string match on the two.

The fact that you still have a second key fingerprint in `~/.ssh/known_hosts` is what means you can still connect - after the hostname matching for Teleport fails, it just falls back to using the fingerprint you already trusted. You can delete the one stored against the IP once you've fixed the `HostKeyAlias` and it should still work.

**Reply by lamhoangtung** [reply]
Thanks @webvictim. I changed `HostKeyAlias` to `techainer-rtx3090-1.teleport` and it is working as expected now

**Comment by lamhoangtung** [comment]
I'm on MacOS 11.1 + zsh @webvictim . The login script seems to fail for the login with:

```console
inappropriate ioctl for device
ERROR: inappropriate ioctl for device
```

Full log with `-vv`:
```console
OpenSSH_8.1p1, LibreSSL 2.7.3
debug1: Reading configuration data /Users/linus/.ssh/config
debug1: /Users/linus/.ssh/config line 11: Applying options for techainer-rtx3090-1
debug1: /Users/linus/.ssh/config line 142: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 47: Applying options for *
debug1: Executing proxy command: exec sh -c '~/.ssh/teleport/login.sh tele.techainer.com linus && ssh tele.techainer.com -s proxy:techainer-rtx3090-1:3022'
debug1: identity file /Users/linus/.tsh/keys/tele.techainer.com/linus type 0
debug1: identity file /Users/linus/.tsh/keys/tele.techainer.com/linus-cert type 4
debug1: Local version string SSH-2.0-OpenSSH_8.1
debug1: kex_exchange_identification: banner line 0: /usr/bin/ssh-keygen
debug1: kex_exchange_identification: banner line 1: Enter password for Teleport user linus:
inappropriate ioctl for device
ERROR: inappropriate ioctl for device
```


`tsh login ...` before `ssh ...` still working as expected.

**Reply by lamhoangtung** [reply]
As suggested somewhere on the internet, I had tried `export GPG_TTY=$(tty)` but it does not seems to resolve the problem

**Reply by webvictim** [reply]
Ah, this is interesting. The reason is because Teleport is prompting you for a password via the login command:

```
kex_exchange_identification: banner line 1: Enter password for Teleport user linus:
```

I personally use Github authentication (which opens a browser window to handle authentication) on the cluster where I run these scripts which is why I haven't noticed this before. This is something I'm not really sure how to work around, as Teleport doesn't currently have any method implemented to send it a password using anything other than the keyboard - this is why you see the `inappropriate ioctl for device`.

If you run `tsh login` manually (and enter your Teleport password) to get valid credentials before trying to connect then it should work.

**Reply by lamhoangtung** [reply]
Thanks @webvictim. Can we open an issue to speak up for password login method other than the keyboard ... ?

**Reply by webvictim** [reply]
Absolutely, feel free to open an issue with a feature request if you'd like. Glad it's working for now!

**Reply by lamhoangtung** [reply]
Thanks @webvictim for the help. For now, I modified the `login.sh` script a little bit like this:

```bash
#!/bin/bash
TSH=$(which tsh)
if [ ! -f ${TSH} ]; then
    echo "Cannot find tsh at '${TSH}'" > /dev/tty
    exit 2
fi

if ! command -v ssh-keygen; then
    echo "Cannot find ssh-keygen" > /dev/tty
    exit 2
fi

if [[ "$1" == "" ]]; then
    echo "Usage: $(basename $0) <proxy> <username>" > /dev/tty
    exit 1
fi
PROXY=$1

if [[ "$2" == "" ]]; then
    echo "Usage: $(basename $0) <proxy> <username>" > /dev/tty
    exit 2
fi
USERNAME=$2

if [ -f ~/.tsh/keys/${PROXY}/${USERNAME} ]; then
    # the date command is different on OS X, of course
    if [[ $OSTYPE == darwin* ]]; then
        EXPIRY_TIME=$(date -j -f '%Y-%m-%dT%H:%M:%S' `ssh-keygen -Lf ~/.tsh/keys/${PROXY}/${USERNAME}-cert.pub | grep Valid | cut -d' ' -f13` +%s)
    elif [[ $OSTYPE == linux* ]]; then
        EXPIRY_TIME=$(date -d `ssh-keygen -Lf ~/.tsh/keys/${PROXY}/${USERNAME}-cert.pub | grep Valid | cut -d' ' -f13` +%s)
    else 
        echo "Unsupported OS '${OSTYPE}'" > /dev/tty
        exit 2
    fi

    NOW=$(date +%s)
    if [ ${NOW} -lt ${EXPIRY_TIME} ]; then
        exit 0
    fi
fi

LOGIN_STATUS=$(${TSH} status)
if [ -z "$LOGIN_STATUS" ]
then
      echo "Please login first using tsh login ... and try again" > /dev/tty
      exit 2
else
      echo "Done login script"
fi

# ${TSH} login --proxy=${PROXY} --user=${USERNAME}
```

So that I can see the message `Please login first using tsh login ... and try again` if the script failed

**Reply by webvictim** [reply]
Awesome - thank you!

**Comment by stefansedich** [comment]
@webvictim just curios, I am setting things up for VSCode access, everything is working fine but I have found I don't need to create the known hosts file with the hosts public key as specified, appears to work fine without performing this step.

Has something changed or am I doing something different with my setup? when I initially ssh I get asked to accept two keys, but after that I am logged in without issue.

**Reply by stefansedich** [reply]
Ah ok maybe I am wrong about it "working", it has issues connecting on subsequent attempts with host identification changed errors.

**Reply by webvictim** [reply]
Yes, if you don’t set the host CA up correctly you’ll get TOFU prompts when your certificates rotate.

**Reply by stefansedich** [reply]
I assume there is no more automated way to get the host public key than to use `tctl` as specified above? in my instance I am bakling the public keys into our `setup.sh` script which works great but if I can avoid that it would be great.

**Reply by webvictim** [reply]
@Joerger Do we have any public API endpoints for this?

**Reply by Joerger** [reply]
We don't have a public API endpoint for that yet, but it could be added. If its needed, feel free to make an [issue for `GetCertAuthority`](https://github.com/gravitational/teleport/issues/7125) so that it can be prioritized.

**Comment by lamhoangtung** [comment]
For Teleport 6.2+, if you are getting:
```console
teleport_user@teleport.example.com: Permission denied (publickey).
```

You might need to check if an ssh-agent are running (Arch does not have ssh-agent enable by default for example) and re-login into your teleport account.

Just spend a few hours with @santapo debugging this. Hope it can help someone ;)

**Reply by webvictim** [reply]
Yes, you're absolutely right. I noticed this myself yesterday.

I've just updated the original post to add details on `CertificateFile` which is needed for this to work without using `ssh-agent`.

**Reply by lamhoangtung** [reply]
Thanks @webvictim

**Comment by slavitch** [comment]
Hello, I've enabled x11-forwarding in my Teleport roles. I get this error when trying to connect using ssh -Y or ssh -X
```
node doesn't support request type 'x11-req'
```

How do I forward X11 from a teleport node?

**Reply by webvictim** [reply]
Teleport nodes don't support X forwarding directly. You'll need to connect to `sshd` on port 22.

See this section of the docs on how to configure `sshd` to trust Teleport-issued certificates: https://goteleport.com/docs/server-access/guides/openssh/#set-up-openssh-recording-proxy-mode
